src/contracts/batching/mod.rs

```diff
@@ -0,0 +1,136 @@
+mod abi;
+
+use self::abi::BatchingContract as ContractAbi;
+use crate::{
+    contracts::{legacy::MemberAddedEvent, IdentityManager, Options},
+    ethereum::{Ethereum, EventError, Log, ProviderStack, TxError},
+    tx_sitter::Sitter,
+};
+use async_trait::async_trait;
+use ethers::{
+    providers::Middleware,
+    types::{TransactionReceipt, U256},
+};
+use futures::Stream;
+use semaphore::Field;
+use std::pin::Pin;
+use tracing::{error, info, instrument};
+
+// TODO [Ara] Remove the allows.
+/// A structure representing the interface to the batch-based identity manager
+/// contract.
+pub struct Contract {
+    #[allow(dead_code)]
+    ethereum: Ethereum,
+    #[allow(dead_code)]
+    sitter:   Sitter,
+    #[allow(dead_code)]
+    abi:      ContractAbi<ProviderStack>,
+}
+
+#[async_trait]
+impl IdentityManager for Contract {
+    #[instrument(level = "debug", skip_all)]
+    async fn new(options: Options, ethereum: Ethereum) -> anyhow::Result<Self>
+    where
+        Self: Sized,
+    {
+        // Check that there is code deployed at the target address.
+        let address = options.identity_manager_address;
+        let code = ethereum.provider().get_code(address, None).await?;
+        if code.as_ref().is_empty() {
+            error!(
+                ?address,
+                "No contract code is deployed at the provided address."
+            )
+        }
+
+        // Connect to the running batching contract.
+        let abi = ContractAbi::new(
+            options.identity_manager_address,
+            ethereum.provider().clone(),
+        );
+
+        let owner = abi.owner().call().await?;
+        if owner != ethereum.address() {
+            error!(?owner, signer = ?ethereum.address(), "Signer is not the owner of the identity manager contract.");
+            panic!("Cannot currently continue in read-only mode.")
+        }
+        info!(
+            ?address,
+            ?owner,
+            "Connected to the WorldID Identity Manager"
+        );
+
+        let sitter = Sitter::new(ethereum.clone()).await?;
+
+        let identity_manager = Self {
+            ethereum,
+            sitter,
+            abi,
+        };
+
+        Ok(identity_manager)
+    }
+
+    fn tree_depth(&self) -> usize {
+        // TODO [Ara] Work out what to do for this. There is no sane answer for a
+        //   contract that can work with arbitrary verifiers.
+        todo!()
```

Comment:
This could be a regular config option provided on startup via `clap`. No action for now.